pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,511 @@
+	} else {
+		for _, ds := range b.Datasets {
+			meta.AddTenant(ds.TenantId)
+		}
+	}
+
+	return meta, nil
+}
+
+// FindBlock tries to retrieve an existing block from the index. It will load the corresponding partition if it is not
+// already loaded. Returns nil if the block cannot be found.
+func (i *Index) FindBlock(shardNum uint32, tenant string, blockId string) *metastorev1.BlockMeta {
+	// first try the currently mapped partition
+	key := i.CreatePartitionKey(blockId)
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	b := i.findBlockInPartition(key, shardNum, tenant, blockId)
+	if b != nil {
+		return b
+	}
+
+	// try other partitions that could contain the block
+	t := ulid.Time(ulid.MustParse(blockId).Time()).UTC().UnixMilli()
+	for _, p := range i.allPartitions {
+		if p.contains(t) {
+			b := i.findBlockInPartition(p.Key, shardNum, tenant, blockId)
+			if b != nil {
+				return b
+			}
+		}
+	}
+	return nil
+}
+
+func (i *Index) findBlockInPartition(key PartitionKey, shard uint32, tenant string, blockId string) *metastorev1.BlockMeta {
+	meta := i.FindPartitionMeta(key)
+	if meta == nil {
+		return nil
+	}
+
+	p, err := i.getOrLoadPartition(meta)
+	if err != nil {
+		return nil
+	}
+
+	p.shardsMu.Lock()
+	defer p.shardsMu.Unlock()
+	s, _ := p.shards[shard]
+	if s == nil {
+		return nil
+	}
+
+	s.tenantsMu.Lock()
+	defer s.tenantsMu.Unlock()
+	t, _ := s.tenants[tenant]
+	if t == nil {
+		return nil
+	}
+
+	t.blocksMu.Lock()
+	defer t.blocksMu.Unlock()
+	b, _ := t.blocks[blockId]
+
+	return b
+}
+
+// FindBlocksInRange retrieves all blocks that might contain data for the given time range and tenants.
+//
+// It is not enough to scan for partition keys that fall in the given time interval. Partitions are built on top of
+// block identifiers which refer to the moment a block was created and not to the timestamps of the profiles contained
+// within the block (min_time, max_time). This method works around this by including blocks from adjacent partitions.
+//
+// FIXME aleks-p: A large query could cause a large number of partitions to be loaded into memory
+//   - consider loading partitions in parallel
+//   - consider capping the number of loaded partitions
+func (i *Index) FindBlocksInRange(start, end int64, tenants map[string]struct{}) ([]*metastorev1.BlockMeta, error) {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	blocks := make([]*metastorev1.BlockMeta, 0)
+
+	firstPartitionIdx, lastPartitionIdx := -1, -1
+	for idx, meta := range i.allPartitions {
+		if meta.overlaps(start, end) {
+			if firstPartitionIdx == -1 {
+				firstPartitionIdx = idx
+			}
+			p, err := i.getOrLoadPartition(meta)
+			if err != nil {
+				level.Error(i.logger).Log("msg", "error loading partition", "key", meta.Key, "err", err)
+				return nil, err
+			}
+			tenantBlocks := i.collectTenantBlocks(p, tenants)
+			blocks = append(blocks, tenantBlocks...)
+		} else if firstPartitionIdx != -1 {
+			lastPartitionIdx = idx - 1
+		}
+	}
```

Comment:
Use of the mutexes is inconsistent: we do not lock access to shards and tenants in `getOrLoadPartition` (it is safe because we took global `partitionMu`) but do acquire locks in `collectTenantBlocks`; note that the locks are redundant there